modules/README.md

```diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+interface ISignatureValidatorHooks {
+    /**
+     * @param account Address of the account for which signature is being validated
+     * @param validator Address of the validator contract to be used for signature validation. This address will be account address in case of default signature validation flow is used.
+     * @param payload The payload provided for the validation
+     * @return result bytes containing the result
+     */
+    function preValidationHook(address account, address validator, bytes calldata payload) external returns (bytes memory result);
+
+    /**
+     * @param account Address of the account for which signature is being validated
+     * @param preValidationData Data returned by preValidationHook
+     * @return result bytes containing the result
+     */
+    function postValidationHook(address account, bytes calldata preValidationData) external returns (bytes memory result);
+}
+```
+
+### Signature validator manager
+
+A signature validator manager must implement the following interface.
+
+```solidity
+interface ISafeProtocolSignatureValidatorManager {
+
+    /**
+     * @notice A view function that the Manager will call when an account has enabled this contract as a function handler in the Manager for function isSignatureValid(bytes32,bytes)
+     * @param account Address of the account whose signature validator is to be used
+     * @param sender Address requesting signature validation
+     * @param data Calldata containing the 4 bytes function selector, 4 bytes signature selector, and one of the below:
+     *              encoded bytes containing bytes32 domain separator, bytes32 messageHash and bytes32 signatures
+     *              OR
+     *              encoded bytes containing bytes32 messageHash and bytes signatures
+     */
+    function handle(
+        address account,
+        address sender,
+        uint256 /* value */,
+        bytes calldata data
+    ) external view override returns (bytes memory);
+
+    /**
+     * @param domain bytes32 containing the domain for which Signature Validator contract should be used
+     * @param signatureValidatorContract Address of the Signature Validator Contract implementing ISafeProtocolSignatureValidator interface
+     */
+    function setSignatureValidator(bytes32 domain, address signatureValidatorContract) external;
+
+    /**
+     * @param signatureValidatorHooksContract Address of the contract to be used as Hooks for Signature Validator implementing ISignatureValidatorHook interface
+     */
+    function setSignatureValidatorHooks(address signatureValidatorHooksContract) external;
+}
+```
+
+### Security considerations
+
+A malicious actor can change the routing of signature validation flow by changing the 4 bytes of signature selector.
+Under such cases the signature validation should fail and revert the transaction. To prove that signature validation fails due to changed routing, formal verification can be used. This would be tackled in future.
```

Comment:
Please create an issue for this.

Reply:
https://github.com/safe-global/safe-core-protocol-specs/issues/47